### DISCUSSION TITLE
2.6 needs a Game version like 2.5 had

### DISCUSSION THREAD

**Discussion by TheDog-GH** [triplea-game/triplea#12441] [General]
@asvitkine 
EDIT: Game Version ~~Map Version:~~ is missing on the Selected Map panel.

Screen of 2.5
![image](https://github.com/triplea-game/triplea/assets/71022517/a7811c8a-f94c-42e5-8fdc-5f617b581975)

.
Screen 2.6
![image](https://github.com/triplea-game/triplea/assets/71022517/e812e5d9-7227-4c54-8d35-2ae7ee5be3e1)

**Comment by Cernelius** [comment]
Actually, this was a change that @DanVanAtta made deliberately if I recall correctly.

I guess the labels "Problem" and "Regression" should be removed, but I'll let someone else do it.

By the way, it is wrong to say "Map Version": it should have been called "Game Version".

**Comment by DanVanAtta** [comment]
I vaguely recall the reasoning for removal. The gist is `Map Version` is not used for anything. If you're sophisticated enough to switch map versions & to know which version you are using, you don't really need the UI to tell you.

**Comment by DanVanAtta** [comment]
`By the way, it is wrong to say "Map Version": it should have been called "Game Version".`

Map version is probably more correct than 'game version.' The terminology is pedantic and confused. A 'Game' is an instance of a scenario being played on a map. 

EG: 

Game name: "Player vs AI  Game-1"
Game date: 2024-05-12 12:00
Game round: 1
Map Name: "generic WWII map"
Scenario: "1941"

Considering that maps are bundled with scenarios, they are essentially 1:1. Which means "scenario" & "map" are pretty interchangeable. Since we are talking about "scenario" here, "map" is arguably the more correct (but still incorrect). 

Unless someone is really mucking around in the XML, I don't see the 'map/game' distinction as particularly helpful. The logic above would have us do better to use the word 'map' than 'game' when it comes to user facing terminology.

**Comment by TheDog-GH** [comment]
I have a novice computer user and I cannot ask him what version of the map/xml he has, so this would have been ideal.

My answer is to get them to remove and reinstall the map & hope they do it correctly.


Its a nice to have, so no biggie.

**Comment by beelee1** [comment]
@TheDog-GH yea as a workaround, I had to add a notification at game start saying which version it is, so players know they have the same one. 
You would have to have them open up the xml otherwise. Obviously not ideal for a lot of users.

**Reply by DanVanAtta** [reply]
I just groked this statement a bit more.. cc: @TheDog-GH   

TripleA is was simply never designed to have multiple versions of the same scenario. This is a fundamental limitation. It's how it was since the system was built circa 2003~2005. The engine was originally coded to simply find a map folder that matched whatever the map name was.  Multiplayer was added after singleplayer, the consideration for multiple versions probably came into play, but only once the sub-system for loading maps had already been built. That subsystem was never revised to fully support the concept of multiple map versions. Thus, the engine assumes that a map is solely identified by its name, and that identity is provided by the name of the map folder.

FWIW, over time there were some updates to make that map-name to folder matching more flexible. 2.6 changes this and moves the source-of-truth for the identify of a map to a configuration file within that map (maps.yml).  Otherwise the map-loading system is largely unchanged since how it was originally built (and at that time, it was built with just single-player in mind).

**So, if versions are really important - the engine would prefer for you to release completely new mod files. EG: 'my-mod-v1.xml', 'my-mod-v2.xml'. Alternatively & where possible, keep the updates backward compatible in which case it would not matter.**

Making backward incompatible updates to a map puts the system into something of an unknown state. 


>  I had to add a notification at game start saying which version it is, so players know they have the same one.

If it's the case that there was an update and all players should update to the latest version of that map, this is what the version value in `triplea_maps.yaml` is for. Increment that value and players will be prompted to update their map. If a backward incompatible change has been made, it's expected for that version value to be incremented - perhaps this was not done? (at the same time, backward incompatible changes are strongly discouraged as the game engine does not cleanly support the use-case).

The version in 'triplea-maps.yaml' is otherwise expected to be incremented rarely when there are a sufficient number of nice-to-have updates that are relatively finalized that players would do well (but are not required) to update their maps.

At the same time, updating map versions on a players system has implications for save games, which makes the whole thing tricky. I don't know of any exact messaging, but the way it is setup, it's expected for players to just know when they need to wait on updating a map to finish any save games.

-------------------------

In way, the actual version that matters is what is in maps-index. Which is the value that was previously in `maps.xml` and now in `triplea_maps.yml`.   Arguably that should be the version value that was ever displayed as the 'map-version', and arguably if we are going to restore the feature then the 'triplea_maps.yml' version would be the best one to show, for these reasons:
  - all maps a version value in 'triplea_maps.yml'
  - that version value does something
  - that version value is intended to represent compatible/incompatible updates
  -  the semantics of that version value are simple & consistent, which mean they can be readily interpreted by the engine

**Comment by Cernelius** [comment]
> Considering that maps are bundled with scenarios, they are essentially 1:1.

If by 1:1 you are referring to the maps featuring a single game, this is simply not true in general whether or not these maps might be the majority (not sure). How about looking at the top 3 most popular maps in the lobby: they are World At War, Global and Revised.

What of these maps is a single-game (or single-scenario) map?

None.

The "world_at_war" map features 2 games, namely "World At War" and "WAW 1940".

The "world_war_ii_global" map features 8 (!!!) games, namely "World War II Global 1940 Canadian Mod with Combat Move First", "Ozteas 1941 Global Setup", "World War II Global 1940 Original", "World War II Global 1940 2nd Edition", "World War II Global 1940 2nd Edition with Combat Move First", "World War II Global 1940 Balanced Mod3", "World War II Global 1940 Balanced Mod3 with Combat Move First" and "World War II Global 1942 2nd Edition".

The "world_war_ii_revised" map features 2 games, namely "World War II Revised LHTR" and "World War II Revised".

By the way, I think that the "world_war_ii_revised" map should actually feature 7 games in total because 5 more are inside the "world_war_ii_revised_variations" map (which is just a skin clone of the "world_war_ii_revised" map as far as I know) and should rather be bundled inside the former (If two maps are the same map, they should be the same map to avoid needless and potentially harmful duplication.).

However, I tend to agree that what is currently called a "game" may be called a "scenario" if preferred (though I prefer the term game but that may be just due to the fact that I'm into TripleA since a long time and that has been the name for me: the folder where the XML are located is called "games", not "scenarios", for example).

---

However, I see your point that every "game" which I play of the same "game" is a game in itself, so (for example), if I play 10 games of the "270BC Wars" game, I can see those as 10 different games of the same game, which does sound self-contradictory. I think that here there is indeed the same term used for two different things, which is not great, but this is really a language problem (Also if I play 100 games of "chess", I can say that I've played 100 games of the same game and everyone would understand what I'm talking about.), and I think that it is really not a big deal: those 10 played games are further split into as many game-saves as taken in the course of each of the games, and we are not operating on game-saves (we just look at them).

Notice, moreover, that the splitting of a same game into multiple games does not only happen for the former (which is merely the common starting point of each of the latter) as it also happens whenever I decide to reload. For example, if I play a game until achieving victory on round 10 and thereafter I decide to reload and re-play from round 5, I now have two different games of the same game which share all game-saves until round 5 but have different game-saves thereafter. On this regard, the reloading point at round 5 is conceptually the same thing as the original game, as that can be regarded as a game-save taken after having played nothing of the game, thus justifying this apparent contradiction of calling two different things with the same name on the basis that they can be actually seen as being the same thing!

I would agree about opening in forum a poll on whether or not to rename the "games" as "scenarios", but I would vote to keep calling them "games".

However, if "games" would be changed to "scenarios" in the UI, then you would be left with the fact that "scenarios" are called "games" in several non-UI places (like the "games" folder or in the "map.yml"), which is not a big deal in general but would mean that every new developer or power-user will have to be told that "scenario" and "game" mean the same thing (beside of course the fact that you would need to go through all the extant map-making documentation and update it to the new nomenclature).

**Reply by DanVanAtta** [reply]
> If by 1:1 you are referring to the maps featuring a single game, this is simply not true in general whether or not these maps might be the majority (not sure).

I was incorrect in how I phrased it. What I intended to convey was that from a player's perspective, at a cursory glance, scenarios/maps appear to be 1:1. The fact they are different is something of an engine detail that is heavily exposed to map makers. Exposing this exact terminology and these details to players I don't think is even really necessary. Thus, the intent is to label things more from a users perspective.

I think this speaks a bit to some new ideas from "Domain Driven Design." It was very common in the early days of programming (which is anything more than 5 years ago), for the terminology to start with a coder writing up some code and then exposing those names to the user. Sometimes those names truly match what a user would use, sometimes it does not match. In this case, the person decided to call an XML file a 'game', the map-images (and related stuff like polygons) a map, and the game-in progress is a `GameData` and when saved is a `SaveGame`.

The "Domain Driven Design" thinking takes notes of this and emphasizes to use the player terminology in the code, and to align the names and meanings of the code with what the user end user considers those things to mean. So, if we were building TripleA from scratch today, we would be asking the question: "what do players call this?" & "would players use these words in their day-to-day?", and those questions would be asked a lot.


> What of these maps is a single-game (or single-scenario) map?

There are quite a few that are 1:1. I think a majority might even be. Though, in the pool of popular maps & scenarios, those are probably all modded to hell & back (figuratively speaking). So, something of the 80/20 rule applies here. The popular maps are few but heavily modded, most maps are not popular, many have just one XML in them. I believe that to be the case based on some previous analysis I did some years ago, I believe my recollection is likely correct & also that not much has changed since.


> I would agree about opening in forum a poll on whether or not to rename the "games" as "scenarios", but I would vote to keep calling them "games".

The naming of things goes into the code & there has to be willingness to make these updates. To some extent, we are stuck with creating distinctions about what we mean and when (the terminology is pretty deeply baked into the codebase, redoing this gives value for consistency but not much else, and is a ton of work and is unlikely to be fully backward compatible). Thus, when talking in the context of 'map-making', "game" means one thing, when talking to a player "game" means another. Going back to the "DDD" idea on this, the idea is less to open a poll, but just observe the natural language people use.



> However, I see your point that every "game" which I play of the same "game" is a game in itself, so (for example)

I appreciate the consideration and the example. English is just kinda ambiguous in how "a game of Chess" and "a game" can refer to the same thing. I can't really think of an analagous example and somewhat think that "game" in English is just pathological. 

For example:
 
- (speaker-1) We are playing a game. 
- (speaker-2) Which game are you playing?
- (speaker-1) Chess, we just played a game of chess.

It would be interesting for other examples, but typically the answer to "which X were you Y'ing", the answer is usually just "Z", not "a X of Z"

For example, replace "game" with "sport", and "Chess" with "basketball"

- (speaker-1) We are playing a sport. 
- (speaker-2) Which sport are you playing?
- (speaker-1) Basketball, we just played a sport of basketball.



>  which is not a big deal in general but would mean that every new developer or power-user will have to be told that "scenario" and "game" mean the same thing 

I'll point out that this 'translation layer' between developer/player/map-maker is actually pretty deep as it exists. To a developer, players are "nodes".

At the end of the day, "scenario/mod/game" mean the same thing to a player. That could be an open question, whether people commonly believe those to be the same thing (unless they are a developer or a map maker).  I think that is likely to be the case.

"scenario/mod" both mean the same thing to map maker and developers. 


So, it's perhaps just a pity that 'scenario' was not originally chosen. It's a big effort to update the codebase to be consistent, and until then I don't think introducing confusion to map-makers & developers is necessarily worth it. Meanwhile, the language people naturally use is tough to change, and arguably we should not be asking the users to change their terms but instead re-align how we understand those terms instead. In other words, the players tell us what the words mean, not the other way round.

At the end of the day, we should be precise about using the right terms when there is a distinction, but only when that distinction is truly important and secondly to keep in mind that the meaning of a given word is dependent on the audience.

**Comment by Cernelius** [comment]
By the way (off-topic but related), why having "Game Round" showing under "Game Name" also when you load a new game? I think that should show only if you load a game-save.

**Reply by DanVanAtta** [reply]
IIRC the engine allows for a variable round game start.

Otherwise I would be a bit concerned about the UI moving around too much when a new label appears/is hidden, but perhaps it could look clean.

**Reply by Cernelius** [reply]
I actually talked about this with Veq back then (about 10 years ago) when this (kinda strange) feature was introduced (being curious what was the point of allowing us to have a starting round other than 1), and he told me that was just for supporting the XML export (of games in progress), and that no map-makers are supposed to have games (in the repository) starting at any round other than 1. Regardless, there are a few maps each having one or more games with a starting round different from 1. I actually think it's not unreasonable in the moment a game is actually a modification of an original game made in a way to present something like a pseudo-game-save, that is starting at a point as if the game was played until that point. The prominent examples are all games of Age of Tribes beside the Primeval scenario (which is substantially the "main" game/scenario of the map).

All this said, I already knew that round was not necessarily equal to 1 in every case, but I would anyway avoid having such round showing up when loading a new game (not a game-save): the round in a new game is given in the game file, so showing it up is just like showing whatever other things (like the number of players or some property, like if V2 is true or false), and it makes really no sense to show this specific thing in the moment it has no importance.

---

>Otherwise I would be a bit concerned about the UI moving around too much when a new label appears/is hidden, but perhaps it could look clean.

Rather than deleting the line, I would have it as
```
Game Round: -
```

That should clarify that nothing of the game has been played yet, so there is no actually "current" game round.

Moreover, the "Game Round" (left only for game-saves) could be moved below the "Loaded Savegame:" line, so (when you are starting a new game) you would just see:

```
Loaded Savegame: -
Game Round: -
```

or both just absent.

---

On this regard, if there is anything at all (beside the name and the version) which I would show from the game-file (XML), it is the total number of zones in the map. Beside much using zones in hacky ways (which is very rare), that would give the user a pretty consistent idea about what kind of map it is looking at in term of size.

Look at the "Total Areas" colum of this for example:
https://axisandallies.fandom.com/wiki/Category:TripleA_Maps

For example, Revised is 143, Global is 333 and World At War is 538. That way, users should be able to have an idea about the map size before trying out the game, and it would be an interesting information to have.

**Reply by DanVanAtta** [reply]
I think those details are probably best in the map/game selection rather than the main launcher window.

**Comment by DanVanAtta** [comment]
@Cernelius  / @TheDog-GH  / @beelee1  please note I converted this to a 'discussion'  instead. Since we are getting off-topic, the discussion format let's each individual topic be more cleanly discussed.

**Reply by Cernelius** [reply]
I think it was really a discussion since the start anyway. How can it be a problem if it was an intended change which was made as it was supposed to be made?

**Reply by DanVanAtta** [reply]
Indeed, I just wanted to point out the format of the discussion is updated. It took me a while to realize the difference github created in their 'Discussion' vs 'Issue' type. 'Discussion' format is nice for when a conversation has "multiple prongs" vs a single focused topic (EG: software bug) - it's just a different format, still a discussion.

**Comment by TheDog-GH** [comment]
So I would like a version of the map/xml whatever you want to call it, under the game name as a map maker even I get confused when I launch the wrong version of one of my games/maps.

But more importantly I felt I could not support a player, because they could not tell me what version they were running.

**Reply by DanVanAtta** [reply]
In short, you're fighting against the engines design and the map-maker done messed up and instead of releasing a new XML file, they updated an existing one & broke compatibility. 

The engine has no concept of multiple versions, it assumes all players will always be on the latest version. Thus, (devil's advocate here) per the design of the engine, you should check:
- did the map maker update the version of the map in triplea-maps.yaml to indicate an important update that all players should download the updated map?
- there-after, did the player ignore the "update your map" prompt?
- arguably the map maker should revert any updates that would be of concern and re-issue a new XML file as a new map mod

The above framework I do think is limiting.. If we were to restore map version, I have a concern that we are optimizing for a use-case that we should not encourage. Only due to constraints of the game engine. Specifically, "map version" is not a thing for the game engine. The game engine knows nothing of it, according to the engine, a map is solely identified by its name only. Thus, displaying "map version" on the UI indicates to both players & map-makers that "map version" seemingly even is a thing, but it's not (at least not yet).

Second concern, in large part because of the engine design, the map version rarely changes. It's rarely important. Do we really want to show it to every person every single time they launch TripleA? There are potentially other ways to solve the problem without injecting UI elements that are both almost always the same and almost never used. (Grant it, sometimes when something is used/needed, it is very important in that situation)

Third concern: Arguably the correct version to display for 'maps' version is the value from 'triplea-maps.yaml'.

**Reply by DanVanAtta** [reply]
FWIW, 2.6 moves the master index of maps from `triplea-maps.yaml` to an actual database. This makes tracking things like creation & last update times a lot easier.  To that extent, there would be some opportunity to automatically display those values somewhere during game-selection & map-download.

At the same time, nothing precludes 'version' from being something that is displayed in game notes. Perhaps that is an answer?

**Comment by TheDog-GH** [comment]
Ah, I should of been clearer.
I would like the xml version displayed eg
<info name="1941 Global Command Decision" version="1.165"/>
so just the 1.165

2.6 current Downloads does not know when a new version is available, Im not complaining as 2.6 can now actually download maps and it is only temporary, hopefully. 

So for me it would better if TripleA showed the xml version, but I guess I could put it in game notes.

**Reply by DanVanAtta** [reply]
> 2.6 current Downloads does not know when a new version is available

That might be a misperception. 2.6 routinely indexes the map repositories every 10 or 15 minutes. The database that stores that index keeps track of the last time the repository was updated. If the repository is updated, the database sees that and that information is sent to the client.

It's kinda of a big difference. Any update to the map repositories has the same effect as updating the version value in 'triplea_maps.yaml', every time, automatically. Thus, if any update is pushed to any map repository, without any further action - all users with that map installed will see the "your map is out of date, would you like to update it?"

**Reply by TheDog-GH** [reply]
AFAIK the current 2.6 file download changed by @asvitkine uses the old https://github.com/triplea-game/triplea/blob/master/triplea_maps.yaml the same as 2,5 download mechanism.

**Comment by TheDog-GH** [comment]
@DanVanAtta 
Im very conscious of **your time** for TripleA, so from my perspective this is closed.  You have the best interests of TripleA at heart and I dont need long explanations, just a sentence or two, you dont even need to convince me.  I would prefer your precious time be spent on fixing TripleA.  :)

**Reply by DanVanAtta** [reply]
Appreciate the sentiment. I do get dragged into spending too long communicating. It's hard to be brief and still also correct & coherent : )

I also do not want to just streamroll the conversations at the same time. It is a difficult balance. I appreciate the consideration and your efforts here as well, I want to respect those as much as you respect mine.

**Comment by Cernelius** [comment]
Now that I think about it, when loading a game-file, you have this line:

```
Loaded Savegame: -
```

This is not good because it can mean that either no game-save is loaded or I've loaded a game-save named "-". When no game-save is loaded, either this line should be absent or the "-" should be exchanged for nothing or for a character which is not allowed alone in file-names (at least of Windows) (for example, a white space alone).

**Reply by DanVanAtta** [reply]
~~Is this a proposed update, or is that something from 2.6? I'm slightly confused.~~

*Edit*: yeah, looks like that might be 2.6; I'm not sure if I'm completely happy with that now.

**Reply by DanVanAtta** [reply]
@Cernelius I propose we table this conversation. There's no capacity to make any of the proposed updates, we need to focus instead on other areas that are more pressing before we can come back to this.

**Reply by Cernelius** [reply]
I double checked: this is the same both for 2.5 and 2.6.

I agree that this is very scarcely important, but, just for principle, in the moment in which I can have a game-save called "-", the program should not say the game-save is "-" to mean that no game-save is loaded.
